kong/plugins/zipkin/handler.lua

```diff
@@ -168,23 +188,39 @@ if subsystem == "http" then
 
     ctx.zipkin = {
       request_span = request_span,
-      header_type = header_type,
       proxy_span = nil,
       header_filter_finished = false,
     }
   end
 
 
-  function ZipkinLogHandler:access(conf) -- luacheck: ignore 212
-    local zipkin = get_context(conf, kong.ctx.plugin)
+  local function get_inject_ctx(conf, extract_ctx)
+    local zipkin = get_context(conf, kong.ctx.plugin, extract_ctx)
     local ngx_ctx = ngx.ctx
 
     local access_start =
       ngx_ctx.KONG_ACCESS_START and ngx_ctx.KONG_ACCESS_START * 1000
       or ngx_now_mu()
-    get_or_add_proxy_span(zipkin, access_start)
 
-    propagation.set(conf.header_type, zipkin.header_type, zipkin.proxy_span, conf.default_header_type)
+    local proxy_span = get_or_add_proxy_span(zipkin, access_start)
+
+    local inject_ctx = extract_ctx
+    inject_ctx.trace_id      = proxy_span.trace_id      or inject_ctx.trace_id      or nil
+    inject_ctx.span_id       = proxy_span.span_id       or inject_ctx.span_id       or nil
+    inject_ctx.parent_id     = proxy_span.parent_id     or inject_ctx.parent_id     or nil
+    inject_ctx.should_sample = proxy_span.should_sample or inject_ctx.should_sample or nil
+    inject_ctx.baggage       = proxy_span.baggage       or inject_ctx.baggage       or nil
+    return inject_ctx
+  end
+
+
+  function ZipkinLogHandler:access(conf) -- luacheck: ignore 212
+    propagation.propagate(
+      propagation.get_plugin_params(conf),
+      function(extract_ctx)
+        return get_inject_ctx(conf, extract_ctx)
+      end
```

Comment:
Creating closure for each request is not friendly for the JIT compiler, could you replace it with a local function?

Reply:
done, thanks